### ISSUE TITLE
-g4 option cause a mismatch between wasm and wast files

### ISSUE THREAD

**Issue by z--z---z-----** [WebAssembly/binaryen#1470]
Hello,

This C function:

unsigned int xor( unsigned int a, unsigned int b)
{
    return a^b;
}

When compiled with -g4 and with -O3 (emcc xor.c -O3 -g4 -s WASM=1 -s SIDE_MODULE=1 -o xor.wasm) generates:

wast(optimized):
-------
(func $_xor (; 0 ;) (param $0 i32) (param $1 i32) (result i32)
  (local $2 i32)
  ;;@ xor.c:3:0
  (set_local $2
   (i32.xor
    (get_local $1)
    (get_local $0)
   )
  )
  (get_local $2)
 )
 
wasm(*not* optimized):
--------
(func $func0 (param $var0 i32) (param $var1 i32) (result i32)
    (local $var2 i32)
    get_local $var1
    get_local $var0
    i32.xor
    set_local $var2
    get_local $var2
  )

Without specifying the -g4 option, I get the correctly optimized wasm file.

Please let me know if you need any other information

**Comment by kripken** [comment]
Those two are identical, aren't they (except for the debug info)? If you use wasm-dis from binaryen it's clearer when it's not in the stacky format. Both have one set and 3 gets.

**Comment by z--z---z-----** [comment]
The number of gets and sets are identical, but the sequence of instructions is not the same.
Would it be a confusing debugging experience?

**Comment by binji** [comment]
Yes, the files are the same. The first is using the [folded format](http://webassembly.github.io/spec/core/text/instructions.html#folded-instructions) and the second is using the flat format. Binaryen only supports the folded format.

You can use [wat-desugar](https://github.com/webassembly/wabt#wabt-the-webassembly-binary-toolkit) to convert between the two:

```sh
$ cat test.wat
(func $_xor (; 0 ;) (param $0 i32) (param $1 i32) (result i32)
(local $2 i32)
;;@ xor.c:3:0
(set_local $2
(i32.xor
(get_local $1)
(get_local $0)
)
)
(get_local $2)
)

$ wat-desugar test.wat
(module
  (func $_xor (param $0 i32) (param $1 i32) (result i32)
    (local $2 i32)
    get_local $1
    get_local $0
    i32.xor
    set_local $2
    get_local $2)
  (type (;0;) (func (param i32 i32) (result i32))))

$ bin/wat-desugar test.wat --fold
(module
  (func $_xor (param $0 i32) (param $1 i32) (result i32)
    (local $2 i32)
    (set_local $2
      (i32.xor
        (get_local $1)
        (get_local $0)))
    (get_local $2))
  (type (;0;) (func (param i32 i32) (result i32))))
```
